developer-experience/readme.md

```diff
@@ -0,0 +1,203 @@
+
+The essential tasks are identified below.
+
+- Build - Verify that changes are free of syntax error and compile.
+- Test - Verify that all automated tests pass.
+- Start - Launch end-to-end to simulate execution in a deployed environment.
+- Debug - Attach debugger to started solution, set breakpoints, step through code, and inspect variables.
+
+If effort is invested to make these activities as easy as possible, **the returns on that effort will be multiply the longer the project runs and the larger the team**.
+
+## Defining End-to-End
+
+This document makes several references to running a solution end-to-end (aka E2E). End-to-end for the purposes of this document is scoped to the software that is owned, built, and shipped by the team. Systems owned by other teams or third-party vendors is not within the E2E scope for the purposes of this document.
+
+## Goals
+
+- Maximize the amount of time engineers spend on writing code that fulfills story acceptance and done-done criteria.
+- Minimize the amount of time spent manual setup and configuration of tooling
+- Minimize regressions and new defects by making end-to-end testing easy
+
+## Impact
+
+Developer experience can have a significant impact on the efficiency of the day to day execution of the team. A positive experience can pay dividends throughout the lifetime of the project; especially as new developers join the team.
+
+- Increased Velocity - Team spends less time on non-value-add activities such as dev/local environment setup, waiting on remote environments to test, and rework (fixing defects).
+- Improved Quality - When its easy to debug and test, developers will do more of it. This will translate to fewer defects being introduced.
+- Easier Onboarding & Adoption - When dev essential tasks are automated, there is less documentation to write and, subsequently, less to read to get started!
+
+**Most importantly, the customer will continue to accrue these benefits long after the code-with engagement.**
+
+## Measures
+
+### Time to First E2E Result (aka F5 Contract)
+
+Assuming a laptop/pc that has never run the solution, how long does it take to setup and run the whole system end-to-end and see a result.
+
+### Time To First Commit
+
+How long does it take to make a change that can be verified/tested locally. A locally verified/tested change is one that passes test cases without introducing regression or breaking changes.
+
+## Participation
+
+Providing a positive developer experience is a team effort. However, certain members can take ownership of different areas to help hold the entire team accountable.
+
+### Tech Lead - Set the bar
+
+The following are examples of how the Tech Lead might set the bar for dev experience
+
+- Determines development environment (IDE, hosting, etc)

```

Comment:
IDE seems like something that developers should be free to choose, but I do see value in sharing common configurations for sensible choices.

I.e., a checkstyle configuration for Intellij can be shared across team members without mandating that all developers should use that IDE

Reply:
I think i can probably reword this to suggest that the team invest in deep integration with a single IDE. Other engineers are free to use something else but, may not be able to gain the benefits of the deep integration set with the default.

Reply:
Adding this note to highlight this is a suggestion, not a mandate

> IDE choice is NOT intended to mandate that all team members must use a single IDE. However, this choice will direct where tight-integration investment will be prioritized. For example, if Visual Studio Code is the **suggested** IDE then, the team would focus on integrating VS code tasks and launch configurations over similar integrations for other IDE's. Team members should still feel free to use their preferred IDE as long as it does not negatively impact the team.